File: engine/validate_dashboard.js
/* validate_dashboard.js — headless render gate for the Aluminum dashboard.
 *
 * Executes the REAL index.html against the REAL data.json in jsdom and fails
 * if the dashboard cannot render. This is the authoritative check: it catches
 * ANY data-shape break generically, because it runs the actual renderer rather
 * than a hand-maintained schema.
 *
 * Why it exists: index.html wraps render() inside the fetch promise chain, so a
 * single .catch() swallows BOTH network errors and rendering exceptions and
 * always prints "Couldn't load data.json". A rendering bug is therefore
 * invisible from the served file alone — data.json can be perfectly valid JSON,
 * serve a 200, and still produce a blank dashboard.
 *
 * Usage: node validate_dashboard.js <index.html> <data.json>
 * Exit 0 = renders clean. Exit 1 = DO NOT PUBLISH.
 */
const fs = require("fs");
const path = require("path");

const [, , IDX, DATA] = process.argv;
if (!IDX || !DATA) { console.error("usage: node validate_dashboard.js <index.html> <data.json>"); process.exit(2); }

let JSDOM;
try { ({ JSDOM } = require("jsdom")); }
catch (e) { console.error("FATAL: jsdom not installed. Run: npm install --prefix /tmp jsdom"); process.exit(2); }

const html = fs.readFileSync(IDX, "utf8");
const raw = fs.readFileSync(DATA, "utf8");

let parsed;
try { parsed = JSON.parse(raw); }
catch (e) { console.error("FAIL: data.json is not valid JSON — " + e.message); process.exit(1); }

const errors = [];

// Surface the exception that index.html's .catch() would otherwise swallow.
const instrumented = html.replace(
  '.catch(()=>{document.getElementById("main").innerHTML=',
  '.catch((__e)=>{__RENDER_ERROR__(__e);document.getElementById("main").innerHTML='
);
if (instrumented === html) {
  console.warn("WARN: could not instrument the .catch() — relying on DOM inspection only.");
}

const dom = new JSDOM(instrumented, {
  runScripts: "dangerously",
  pretendToBeVisual: true,
  beforeParse(w) {
    // Serve the local data.json to the page's fetch().
    w.fetch = () => Promise.resolve({ ok: true, json: () => Promise.resolve(JSON.parse(raw)) });
    w.__RENDER_ERROR__ = (e) => errors.push("RENDER EXCEPTION: " + ((e && (e.stack || e.message)) || String(e)));
    w.addEventListener("error", (ev) => errors.push("UNCAUGHT: " + ev.message));
    // Chart.js is loaded from CDN and is absent offline; stub it so its absence
    // is not mistaken for a data fault.
    w.Chart = function () { return { destroy() {}, update() {} }; };
    w.Chart.register = () => {};
  },
});

setTimeout(() => {
  const doc = dom.window.document;
  const main = doc.getElementById("main");
  const txt = (doc.body && doc.body.textContent) || "";

  // The real failure signal: the .load error card injected into #main.
  // (Do NOT string-match the body — the literal "Couldn't load data.json"
  //  also appears in the inline <script> source, which lives inside <body>.)
  if (main && main.querySelector(".load")) {
    errors.push('ERROR CARD RENDERED: #main contains the "Couldn\'t load data.json" fallback.');
  }
  if (!main || main.innerHTML.trim().length < 500) {
    errors.push("EMPTY RENDER: #main is missing or essentially empty.");
  }

  // Content sentinels — proves the sections that broke today actually populated.
  const D = parsed, O = (parsed.outlook || {});
  const sentinels = [
    ["report date in header", () => new RegExp(String(D.report_date.slice(0, 4))).test(txt)],
    ["so_what headline", () => D.so_what && D.so_what.line && txt.includes(D.so_what.line.slice(0, 40))],
    ["benchmark row", () => D.benchmark && D.benchmark[0] && txt.includes(D.benchmark[0].name)],
    ["sources table", () => D.sources && D.sources[0] && txt.includes(String(D.sources[0][0]).slice(0, 25))],
    ["outlook ai_analysis", () => O.ai_analysis && O.ai_analysis[0] && txt.includes(String(O.ai_analysis[0]).slice(0, 40))],
    ["outlook risks", () => O.risks && O.risks[0] && txt.includes(String(O.risks[0].risk).slice(0, 25))],
    ["outlook catalysts", () => O.catalysts && O.catalysts[0] && txt.includes(String(O.catalysts[0].event).slice(0, 25))],
    ["news", () => D.news && D.news[0] && txt.includes(String(D.news[0].headline).slice(0, 30))],
    ["premiums", () => D.premiums && D.premiums[0] && txt.includes(String(D.premiums[0].name).slice(0, 12))],
    ["elliott wave writeup", () => D.ew && D.ew.long_term && txt.includes(String(D.ew.long_term.writeup).slice(0, 40))],
  ];
  const failed = [];
  for (const [name, fn] of sentinels) {
    let ok = false;
    try { ok = !!fn(); } catch (e) { ok = false; }
    if (!ok) failed.push(name);
  }
  if (failed.length) errors.push("SECTION DID NOT RENDER: " + failed.join(", "));

  if (errors.length) {
    console.error("\n=== DASHBOARD VALIDATION FAILED — DO NOT PUBLISH ===");
    errors.forEach((e) => console.error("  ✗ " + e));
    console.error("");
    process.exit(1);
  }
  console.log("  ✓ renders clean (no exceptions, all " + sentinels.length + " sections populated)");
  process.exit(0);
}, 3000);

File: engine/validate_dashboard.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""validate_dashboard.py — pre-publish gate for the Web_v3 dashboard (snake_case v3 contract).
Run AFTER build_data_json.py, BEFORE git push. Non-zero exit = DO NOT PUBLISH.
    python3 validate_dashboard.py "$ENGDIR" "$SITE"
Asserts every must-have section is present and non-empty (the fatal class is a blank/
placeholder section on the executive dashboard) and enforces freshness (as_of == today).
Prior schema gates kept as *.bak."""
import json, os, sys
ENGDIR = sys.argv[1] if len(sys.argv) > 1 else "."
SITE   = sys.argv[2] if len(sys.argv) > 2 else ENGDIR
DATA   = os.path.join(SITE, "data.json")
INDEX  = os.path.join(SITE, "index.html")
if not os.path.exists(DATA):
    print("FATAL: %s not found — run build_data_json.py first." % DATA); sys.exit(2)
try:
    D = json.load(open(DATA, encoding="utf-8"))
except Exception as ex:
    print("FATAL: data.json is not valid JSON — %s" % ex); sys.exit(1)
f, warns = [], []
def ne(x): return isinstance(x, list) and len(x) > 0
brief=D.get('brief') or {}; lme=D.get('lme') or {}; ew=D.get('elliott_wave') or {}
outl=D.get('outlook') or {}; meta=D.get('meta') or {}
if not brief.get('headline'): f.append("brief.headline empty (hero brief)")
if not ne(brief.get('bullets')): f.append("brief.bullets empty")
if not ne(D.get('kpis')): f.append("kpis empty")
if not ne((lme.get('series') or {}).get('price')): f.append("lme.series.price empty (price chart)")
if not ne(lme.get('benchmarks')): f.append("lme.benchmarks empty")
if not ne((D.get('base_metals') or {}).get('board')): f.append("base_metals.board empty")
if not ne((D.get('premiums') or {}).get('regional')): f.append("premiums.regional empty")
panels = ew.get('panels') or []
if not panels or not ne((panels[0].get('history') or {}).get('price')):
    f.append("elliott_wave first panel history empty")
if not ne((D.get('news') or {}).get('headlines')) and not ne((D.get('peers') or {}).get('earnings')):
    f.append("news AND peers both empty")
if not (ne(outl.get('catalysts')) or ne(outl.get('risks')) or ne(outl.get('consensus'))):
    f.append("outlook empty")
if not ne((D.get('sources') or {}).get('references')): f.append("sources.references empty")
# type sanity
for k in ['kpis']:
    if k in D and not isinstance(D[k], list): f.append("%s must be a LIST" % k)
# freshness
exp = os.environ.get("EXPECT_DATE"); gen = meta.get("as_of")
if exp and gen != exp: f.append("meta.as_of %r != expected %r (stale — refresh did not update)" % (gen, exp))
elif not gen: warns.append("meta.as_of empty — cannot verify freshness")
# shell sanity
if os.path.exists(INDEX):
    html = open(INDEX, encoding="utf-8", errors="ignore").read()
    if 'id="root"' not in html: warns.append('index.html has no <div id="root"> — is it the built shell?')
else:
    warns.append("index.html not in SITE dir (shell committed once; ok if only data.json republished)")
print("── Contract (Web_v3, non-empty critical sections)")
if f:
    for x in f: print("  ✗ " + x)
else:
    print("  ✓ all critical sections present and non-empty")
for w in warns: print("  ! WARN: " + w)
if f:
    print("\n=== VALIDATION FAILED — DO NOT PUBLISH (%d) ===" % len(f)); sys.exit(1)
print("=== VALIDATION PASSED — safe to publish ==="); sys.exit(0)
